### PR TITLE
Refactor upstream handling, prepare PBA offloading and goroutine optimization

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Refactor upstream handling, prepare PBA offloading and goroutine optimization

#### Changes

- Refactored upstream handling; no more switching on gRPC service name with a concrete implementation. This makes things simpler, while not breaking configuration. Also did some necessary decoupling and gateway status handling fix
- Added `GsPba` service and `GatewayUplinkMessage`. The latter is now used in GS because we need the band in upstreams
- Small fix for moving the gc goroutine in each sub-band to the scheduler. This is a 6x goroutine save in EU, for example

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/gatewayserver/scheduling/scheduler.go [hunk 1]
```diff
@@ -193,6 +194,28 @@ func (s *Scheduler) findSubBand(frequency uint64) (*SubBand, error) {
 	return nil, errSubBandNotFound.WithAttributes("frequency", frequency)
 }
 
+func (s *Scheduler) gc(ctx context.Context) error {
```
Comment (KrishnaIyer) [inline pkg/gatewayserver/scheduling/scheduler.go]: nice!

File: pkg/gatewayserver/upstream/ns/ns.go [hunk 2]
```diff
@@ -18,45 +18,44 @@ package ns
...
 
-	"go.thethings.network/lorawan-stack/pkg/component"
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/types"
+	"google.golang.org/grpc"
 )
 
+// Cluster represents the interface the cluster.
```
Comment (htdvisser) [inline pkg/gatewayserver/upstream/ns/ns.go]: Maybe reword this comment

File: pkg/gatewayserver/io/io.go [hunk 3]
```diff
@@ -282,16 +327,23 @@ func (c *Connection) ScheduleDown(path *ttnpb.DownlinkPath, msg *ttnpb.DownlinkM
...
 		if fp == nil {
-			return 0, errFrequencyPlanNotConfigured.WithAttributes("id", request.FrequencyPlanID)
+			// The requested frequency plan is not configured for the gateway. Load the plan and enforce that it's in the same band.
+			fp, err = c.fps.GetByID(fpID)
+			if err != nil {
+				return 0, errFrequencyPlanNotConfigured.WithCause(err).WithAttributes("id", request.FrequencyPlanID)
+			}
+			if fp.BandID != c.bandID {
+				return 0, errFrequencyPlansNotFromSameBand
+			}
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io.go]: @htdvisser @KrishnaIyer FYI we're loading the frequency plan in the hot path if the NS's frequency plan doesn't match that of the gateway.

We have these circumstances with `EU_863_870` vs `EU_863_870_TTN` where we do need to check band. We might need to tune frequency plan store caching (it's already enabled, but implement background refresh so they're immediately available from memory).

File: pkg/gatewayserver/upstream/upstream.go [hunk 4]
```diff
@@ -24,16 +24,14 @@ import (
...
 	GetDevAddrPrefixes() []types.DevAddrPrefix
 	// Setup performs all the preparation necessary to connect the handler to a particular upstream host.
-	Setup() error
+	Setup(context.Context) error
 	// ConnectGateway informs the upstream handler that a particular gateway is connected to the front end.
 	ConnectGateway(context.Context, ttnpb.GatewayIdentifiers, *io.Connection) error
-	// HandleUp handles ttnpb.UplinkMessage.
-	HandleUplink(context.Context, ttnpb.GatewayIdentifiers, ttnpb.EndDeviceIdentifiers, *ttnpb.UplinkMessage) error
+	// HandleUp handles ttnpb.GatewayUplinkMessage.
+	HandleUplink(context.Context, ttnpb.EndDeviceIdentifiers, *ttnpb.GatewayUplinkMessage) error
```
Comment (KrishnaIyer) [inline pkg/gatewayserver/upstream/upstream.go]: This needs to be reverted since v2 Offloading needs the gateway identifiers/
